### PR TITLE
Add optical absorption and Rayleigh interactors

### PR DESCRIPTION
**Pull request by hhollenb**
Add optical absorption and Rayleigh interactors and respective tests to be able to do initial testing of optical models. Also includes an optical `Interaction` class for the results of discrete optical interactions.

New files are placed in the `celeritas::optical` namespace, so the class and file names do not have the usual `Optical` prefix. Both interactors don't depend on any physics or material data, so they're rather simple.

The `optical::TackView` class is just a placeholder for testing, and will need to be changed later for when the optical state and track code is set up. I'm including it now so that `RayleighInteractor`'s function signature will match the usual pattern of passing a track view, and also removes ambiguity of passing two `Real3` arrays to a function.

### CODE REVIEW

File: src/celeritas/optical/Interaction.hh
Comment (sethrj) [inline src/celeritas/optical/Interaction.hh:35]: Even though we'll eventually need it, let's leave out polarization for now as discussed.

File: src/celeritas/optical/Interaction.hh
Comment (sethrj) [inline src/celeritas/optical/Interaction.hh]: I think we can leave out failure too since we're not worried about large showers of secondaries.

File: src/celeritas/optical/model/RayleighInteractor.hh
Comment (sethrj) [inline src/celeritas/optical/model/RayleighInteractor.hh]:
```suggestion
 * Rayliegh interaction.
```

Wait, so Rayleigh interaction is just isotropic in the exiting direction if polarization isn't considered?

File: test/celeritas/optical/Rayleigh.test.cc
Comment (sethrj) [inline test/celeritas/optical/Rayleigh.test.cc:23]: This old test base (again, one of the first ones we wrote in celeritas) is probably not suitable for optical physics. I think we could potentially use the optical test base...

File: test/celeritas/optical/Rayleigh.test.cc
Comment (sethrj) [inline test/celeritas/optical/Rayleigh.test.cc:82]: Fancy! First use of the digits separator in our code 😄 we should consider doing this for the other large integers in use...

File: src/celeritas/optical/model/RayleighInteractor.hh
Comment (hhollenb) [reply src/celeritas/optical/model/RayleighInteractor.hh]: Geant4 uses the Rayleigh scattering model from: G.O. Depaola / Nuclear Instruments and Methods in Physics Research A 512 (2003) 619–630

It treats molecules as isotropic, and the only angular dependence in the differential XS is the angle between the old and new polarizations. So without polarizatoins, Rayleigh will just be isotropic scattering. Anisotropic scattering would require a different model (e.g. https://arxiv.org/pdf/2202.12156 ) but would still probably need polarizations.

File: test/celeritas/optical/Rayleigh.test.cc
Comment (hhollenb) [reply test/celeritas/optical/Rayleigh.test.cc:23]: I'll admit I was being a bit lazy with this one and wanted to get tests out and working -- I'll throw together an optical interactor test base

File: src/celeritas/optical/Interaction.hh
Comment (whokion) [inline src/celeritas/optical/Interaction.hh]:
```suggestion
 * photon is only changed through absorption re-emission processes.
```

File: src/celeritas/optical/Interaction.hh
Comment (whokion) [reply src/celeritas/optical/Interaction.hh]: For a future reference, `reemitted` may be added for the wavelength swift interaction.

File: src/celeritas/optical/model/AbsorptionInteractor.hh
Comment (whokion) [inline src/celeritas/optical/model/AbsorptionInteractor.hh]: We shortly discussed before, but should the energy of absorbed photon be deposited locally (which may provide, at least, MC truth like information inside radiators as well as in the sensitive detectors such as PMT or SiPM) -  of course, the energy deposition can be done elsewhere (at the high level), but followed our standard action for physics interactions.

File: src/celeritas/optical/Interaction.hh
Comment (whokion) [inline src/celeritas/optical/Interaction.hh]: Not sure whether this statement is correct as absorption and WLS are not elastic.  Probably, do you mean that `all photon interactions are discrete`?

File: src/celeritas/optical/Interaction.hh
Comment (whokion) [reply src/celeritas/optical/Interaction.hh:35]: Have we also decided not to allocate any storage for secondaries (which should be fine for now as we do not have WLS yet)?  Just wondering though.

File: src/celeritas/optical/model/AbsorptionInteractor.hh
Comment (whokion) [inline src/celeritas/optical/model/AbsorptionInteractor.hh]: Should we create a separate directory for interactors (i.e, option/interactor )?

File: test/celeritas/optical/Rayleigh.test.cc
Comment (whokion) [inline test/celeritas/optical/Rayleigh.test.cc]: A good additional test is the product of the polarization vector and the direction.

**Review by whokion — APPROVED** [review]
Very good start.  LGTM with several minor comments!

File: src/celeritas/optical/Interaction.hh
Comment (sethrj) [reply src/celeritas/optical/Interaction.hh:35]: Correct, the first iteration let's go for absolute minimum functionality (as long as it doesn't prevent us from adding the new capabilities going forward).

File: src/celeritas/optical/model/AbsorptionInteractor.hh
Comment (hhollenb) [reply src/celeritas/optical/model/AbsorptionInteractor.hh]: The interaction result here will just be the proposed change to the photon track. We definitely need a way to say that the photon has been absorbed because the track will need to be killed. Whether or not that will perform a local energy deposition will be at a higher level where the `Interaction` result gets applied to each track. 

As for recording the actual energy deposit, I've currently written in the assumption that the energy of optical photons never changes, so when an absorption result is returned, the higher level loop should use the photon's current energy (if we decide to deposit it).

File: src/celeritas/optical/Interaction.hh
Comment (hhollenb) [reply src/celeritas/optical/Interaction.hh]: You're very right, I may have been too caught up in thinking about Rayleigh scattering at the time...

File: test/celeritas/optical/Rayleigh.test.cc
Comment (hhollenb) [reply test/celeritas/optical/Rayleigh.test.cc]: The sanity checks should always check that the polarization and direction are perpendicular for each interaction. Do you mean the initial direction - final polarization?

File: src/celeritas/optical/model/AbsorptionInteractor.hh
Comment (whokion) [reply src/celeritas/optical/model/AbsorptionInteractor.hh]: Yes, we can assume that the energy of optical photons is constant along a step. As long as the photon energy is deposited somewhere in the higher level, it should be okay.  Extended to WLS, the energy is not conversed (as some of energy will be dissipated through phonons or heat during absorption and reemission), so the difference in energy may be deposited locally even though this delta energy is not currently deposited in Geant4.  So, I was thinking that any energy deposition may be handled inside physics interactors consistently (if we support fine detailed optical simulation inside layers of a fancy photo-tube, like SIPM).  Either way probably be okay for now.

File: test/celeritas/optical/Rayleigh.test.cc
Comment (whokion) [reply test/celeritas/optical/Rayleigh.test.cc]: I overlooked the sanity check, which may also include  `dot_product(final direction, final polarization) = 0`

File: src/celeritas/optical/model/AbsorptionInteractor.hh
Comment (hhollenb) [reply src/celeritas/optical/model/AbsorptionInteractor.hh]: Thanks, that's good to know.

If we have access to the secondaries from the interaction, we should be able to calculate the energy deposition at a higher level, and so we don't have to check energy conservation across every interactor. I don't have a strong opinion either way at the moment.

We may also want to consider adding WLS now since it seems to have more complications to keep track of than Rayleigh or absorption.

Comment (whokion) [reply src/celeritas/optical/model/AbsorptionInteractor.hh]: Good. It would be better to merge this MR first so that we move to integrate and move to the rest of the optical photon loop.  WLS implementation needs more brain-storming for how to support multiple time structures which is poorly designed in Geant4 (i.e., WLS, WLS2 which are just same codes, but with different reemission functions, which can be combined within one model).

File: src/celeritas/optical/Interaction.hh
Comment (amandalund) [reply src/celeritas/optical/Interaction.hh:35]: We don't have to decide in this PR, but if the track initialization stuff will be similar to what's in our main stepping loop it's possible it will be easier to include secondaries from the start rather than removing that logic and adding it back in later.

Comment (sethrj) [reply src/celeritas/optical/Interaction.hh:35]: Perhaps... looking at the code though, the "locate tracks" executor uses core params and physics step view, and the initialization-from-secondaries is likewise complicated...

File: src/celeritas/optical/interactor/RayleighInteractor.hh
Comment (amandalund) [inline src/celeritas/optical/interactor/RayleighInteractor.hh]: Should this be
```suggestion
        !BernoulliDistribution{ipow<2>(dot_product(new_pol, inc_pol_))}(rng));
```
?

Comment (hhollenb) [reply src/celeritas/optical/interactor/RayleighInteractor.hh]: In terms of Geant4's code, the while loop goes over
```
do {
...
} while (std::pow(cos_theta, 2) < G4UniformRand());
```
so they're accepting with probability $\cos^2 \Theta$. This matches the equations in their references, but differs from the equations in the Geant physics reference manual, which is confusing. I'll reword the documentation to make the sampling method clearer, because the definition of the angles is a bit confusing.

**Comment by sethrj** [comment]
Since the generator/pregenerator code already has polarization in there, I'm taking back what I said about removing it—sorry @hhollenb about the flip-flop and sorry this PR has taken so long. Thanks again for the great work!

**Comment by sethrj** [comment]
@amandalund Could you give a last quick once-over and hit the auto-merge button if it looks good? Thanks!

File: src/celeritas/optical/interactor/RayleighInteractor.hh
Comment (amandalund) [reply src/celeritas/optical/interactor/RayleighInteractor.hh]: Sorry if I'm being dense, but isn't `ipow<2>(dot_product(new_pol, inc_pol_))` equal to $\cos^2 \theta$? Which would mean we're currently _rejecting_ with probability $\cos^2 \theta$ (since `BernoulliDistribution{p_true}(rng)` is equivalent to `u < p_true`)?

File: test/celeritas/optical/Rayleigh.test.cc
Comment (amandalund) [inline test/celeritas/optical/Rayleigh.test.cc]: We could also check the average number of RNG samples.

File: src/celeritas/optical/interactor/RayleighInteractor.hh
Comment (sethrj) [reply src/celeritas/optical/interactor/RayleighInteractor.hh]: Perhaps this is my fault for insisting on using the non intuitive "Bernoulli distribution". Perhaps a helper function that does "reject less than" would be more readable and less error prone. We've already caught one or two of these...

Comment (hhollenb) [reply src/celeritas/optical/interactor/RayleighInteractor.hh]: > Sorry if I'm being dense, but isn't `ipow<2>(dot_product(new_pol, inc_pol_))` equal to cos2⁡θ? Which would mean we're currently _rejecting_ with probability cos2⁡θ (since `BernoulliDistribution{p_true}(rng)` is equivalent to `u < p_true`)?

Nope you're absolutely right, I mixed up that passing the Bernoulli sample was a rejection instead of accepting

File: test/celeritas/optical/Rayleigh.test.cc
Comment (hhollenb) [reply test/celeritas/optical/Rayleigh.test.cc]: Is there a way to query how many times an engine has been sampled, or would I need to make a wrapper?

Comment (amandalund) [reply test/celeritas/optical/Rayleigh.test.cc]: The `DiagnosticRngEngine` used in these tests is a wrapper that does that, so you should be able to call `rng_engine.count()` at the end to get it.

File: src/celeritas/optical/interactor/RayleighInteractor.hh
Comment (amandalund) [reply src/celeritas/optical/interactor/RayleighInteractor.hh]: Yeah we’ve run into this rejection/acceptance confusion with the Bernoulli distribution more than a few times now… you've said before we need a rejection sampler, I think you're right :)

Comment (sethrj) [reply src/celeritas/optical/interactor/RayleighInteractor.hh]: Yep https://github.com/celeritas-project/celeritas/pull/1138#pullrequestreview-1914662560 . I'll put that together right quick...

Comment (sethrj) [reply src/celeritas/optical/interactor/RayleighInteractor.hh]: https://github.com/celeritas-project/celeritas/pull/1354

File: src/celeritas/optical/interactor/RayleighInteractor.hh
Comment (sethrj) [inline src/celeritas/optical/interactor/RayleighInteractor.hh]:
```suggestion
    // Construct interactor from an optical track
    inline CELER_FUNCTION
    RayleighInteractor(TrackView const& particle, Real3 const& direction);

    // Sample an interaction with the given RNG
```

File: src/celeritas/optical/interactor/RayleighInteractor.hh
Comment (sethrj) [inline src/celeritas/optical/interactor/RayleighInteractor.hh]: Should we use a soft equivalence here?

**Review by sethrj — APPROVED** [review]

**Comment by sethrj** [comment]
I went ahead and updated with the rejection sampler to speed this up (and take a little work off your sholders!): @hhollenb the reduced rejection rate here looks correct to me, but I don't know about the updated scattering angles: https://github.com/celeritas-project/celeritas/pull/1317/commits/39c7a511521ce4c1cf6627506318f8244e1bb845

**Comment by hhollenb** [comment]
> I went ahead and updated with the rejection sampler to speed this up (and take a little work off your sholders!): @hhollenb the reduced rejection rate here looks correct to me, but I don't know about the updated scattering angles: [39c7a51](https://github.com/celeritas-project/celeritas/commit/39c7a511521ce4c1cf6627506318f8244e1bb845)

Made a quick plot of the scattered polarization angles (e.g. `dot_product(new_pol, inc_pol_))`) and compared it to the expected $\cos^2\theta$ probability distribution and it looks like they're fairly similar:
rayleigh_dist
There is a little shape difference but I'm not entirely certain if that's within the error bounds or just an artifact of rejection sampling method.

**Comment by sethrj** [comment]
@hhollenb thanks for the plot. It's not surprising the distributions are different since we changed "accept when" to "reject when", as @amandalund suggested. The question is whether the new distribution is correct...

**Pull request merged**
